### ISSUE TITLE
Bug: [no-unnecessary-type-parameters] false positive when used in the implementation of the method

### ISSUE THREAD

**Issue by ST-DDT** [typescript-eslint/typescript-eslint#9828]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.5.4&fileType=.ts&code=KYDwDg9gTgLgBAMwK4DsDGMCWEVxsAZxgB4AVOUfFAEwLiKkxQHMA%2BACgFtgYALCagC44pAJTCGTZnADeAWABQcOGhxE4AQyjM6AXjgByA5rqkA3IuVQeSKLm58BcANSbtBCwoC%2BQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWsRFCAtonyJoqDJCXQO0SODABfELqA&tsconfig=N4XyA&tokens=false

### Repro Code

```TypeScript
export function test<T extends string>(method: T): string {
  const args = '' as T;
  return method + args;
}
```

### ESLint Config

```javascript
{
  "rules": {
    "@typescript-eslint/no-unnecessary-type-parameters": "error"
  }
}
```

### tsconfig

```jsonc
{}
```

### Expected Result

No Error

### Actual Result

Type parameter T is used only once in the function signature.eslint[@typescript-eslint/no-unnecessary-type-parameters](https://typescript-eslint.io/rules/no-unnecessary-type-parameters)

### Additional Info

Original Code: https://github.com/faker-js/faker/blob/cc7715bebded681655b95f5afc368cb250a2eadf/test/support/seeded-runs.ts#L172-L192

The error was first observed in v8.1.0 it does not appear in 7.18.0

The rule seems to ignore any usage of the type inside the actual implementation.

**Comment by JoshuaKGoldberg** [comment]
> The rule seems to ignore any usage of the type inside the actual implementation

Indeed, it does! This is intentional 🙂. See #9735. Uses in the function body shouldn't count. The examples in this issue are both good ones of an unsafe `as` type assertion.

Cross-linking: #9753. This is a good FAQ to have. Thanks for posting!

**Comment by kirkwaiblinger** [comment]
> Cross-linking: #9753. This is a good FAQ to have. 

I think you mean https://github.com/typescript-eslint/typescript-eslint/issues/9680 🙂

**Comment by JoshuaKGoldberg** [comment]
D'oh!

**Comment by eranhirsch** [comment]
Hey all, I have an example of the reported issue here that isn't just using an unsafe `as` and where the suggestion to remove the generic actually hinders the code's type safety. I would love to hear how this could be solved while satisfying the rule without disabling it.

```ts
function lazyImplementation<T, K>(
  keyFunction: (item: T, index: number, data: readonly T[]) => K,
) {
  const set = new Set<K>();
  return (value: T, index: number, data: readonly T[]) => {
    const key = keyFunction(value, index, data);
    if (set.has(key)) {
      return SKIP_ITEM;
    }

    set.add(key);
    return value;
  };
}
```

The gist is that I want to bind the temporary `Set` used to track values to the type the callback function returns. Changing K to `unknown` means that someone can now change the types used when interacting with the Set in ways which break the code silently. e.g.:


```ts
function lazyImplementation<T>(
  keyFunction: (item: T, index: number, data: readonly T[]) => unknown,
) {
  const set = new Set<unknown>();
  return (value: T, index: number, data: readonly T[]) => {
    const key = keyFunction(value, index, data);
    // Oops! now we are checking for the wrong thing :(
    if (set.has(value)) {
      return SKIP_ITEM;
    }

    set.add(key);
    return value;
  };
}
```
